warnings building DPDK with C++17

The dpdk library is using the `register` keyword that got removed in C++17. 

```
[ 18%] Building CXX object lib/dpdk/CMakeFiles/kotekan_dpdk.dir/dpdkCore.cpp.o

In file included from /opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_ether.h:53:0,

                 from /opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_ethdev.h:185,

                 from /data/jenkins/kotekan_rn_c_17/lib/dpdk/dpdkCore.hpp:20,

                 from /data/jenkins/kotekan_rn_c_17/lib/dpdk/dpdkCore.cpp:1:

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder.h: In function ‘uint16_t rte_arch_bswap16(uint16_t)’:

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder.h:56:20: error: ISO C++1z does not allow ‘register’ storage class specifier [-Werror=register]

  register uint16_t x = _x;

                    ^

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder.h: In function ‘uint32_t rte_arch_bswap32(uint32_t)’:

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder.h:71:20: error: ISO C++1z does not allow ‘register’ storage class specifier [-Werror=register]

  register uint32_t x = _x;

                    ^

In file included from /opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder.h:120:0,

                 from /opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_ether.h:53,

                 from /opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_ethdev.h:185,

                 from /data/jenkins/kotekan_rn_c_17/lib/dpdk/dpdkCore.hpp:20,

                 from /data/jenkins/kotekan_rn_c_17/lib/dpdk/dpdkCore.cpp:1:

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder_64.h: In function ‘uint64_t rte_arch_bswap64(uint64_t)’:

/opt/dpdk-stable-16.11.4/x86_64-native-linuxapp-gcc/include/rte_byteorder_64.h:52:20: error: ISO C++1z does not allow ‘register’ storage class specifier [-Werror=register]

  register uint64_t x = _x;

                    ^

```

Needed for #568

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -212,6 +212,10 @@ endif ()
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_GNU_SOURCE -O3 -Wall -Wextra -Werror -march=native -mtune=native -I/opt/rocm/include")
 
+# dpdk uses the register keyword that got removed in C++17 because it wasn't doing anything
+#TODO remove this line when dpdk people fix this: https://mails.dpdk.org/archives/dev/2018-July/107950.html
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-register")
+
 # Warning about missing override is called differently in clang and gcc
 # TODO: enable once we are using a release version of HighFive
 #if (${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -249,7 +249,7 @@ if (${USE_OPENCL})
     install(DIRECTORY lib/opencl/kernels DESTINATION /var/lib/kotekan/opencl )
 endif ()
 set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -lm")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
 
 if (${COMPILE_DOCS})
     add_subdirectory(docs EXCLUDE_FROM_ALL)

--- a/lib/stages/applyGains.cpp
+++ b/lib/stages/applyGains.cpp
@@ -161,7 +161,7 @@ bool applyGains::receive_update(nlohmann::json& json) {
     gainUpdate gain_update = {gain_read, weight_read};
     {
         // Lock mutex exclusively while we update FIFO
-        std::lock_guard<std::shared_timed_mutex> lock(gain_mtx);
+        std::lock_guard<std::shared_mutex> lock(gain_mtx);
         gains_fifo.insert(double_to_ts(new_ts), std::move(gain_update));
     }
     INFO("Updated gains to {:s}.", gtag);
@@ -240,7 +240,7 @@ void applyGains::apply_thread() {
 
         {
             // Request shared lock for reading from FIFO
-            std::shared_lock<std::shared_timed_mutex> lock(gain_mtx);
+            std::shared_lock<std::shared_mutex> lock(gain_mtx);
             gainpair_new = gains_fifo.get_update(double_to_ts(frame_time));
             if (gainpair_new.second == NULL) {
                 FATAL_ERROR("No gains available.\nKilling kotekan");

--- a/lib/stages/applyGains.hpp
+++ b/lib/stages/applyGains.hpp
@@ -94,8 +94,8 @@ private:
     Buffer* out_buf;
 
     /// Mutex to protect access to gains
-    // N.B. `shared_mutex` is only available in C++17
-    std::shared_timed_mutex gain_mtx;
+    // Note: Using shared_mutex together with gcc-5 doesn't work.
+    std::shared_mutex gain_mtx;
 
     /// Timestamp of the current frame
     std::atomic<timespec> ts_frame{{0, 0}};

